openvdb/math/Operators.h

```diff
@@ -1645,7 +1651,9 @@ struct Laplacian<ScaleTranslateMap, DiffScheme>
         ValueType iddz = D2<DiffScheme>::inZ(stencil);
         const Vec3d& invScaleSqr = map.getInvScaleSqr();
         // scale them by the appropriate 1/dx^2, 1/dy^2, 1/dz^2 and sum
-        return ValueType(iddx * invScaleSqr[0] + iddy * invScaleSqr[1] + iddz * invScaleSqr[2]);
+        return iddx * ValueType(invScaleSqr[0]) +
+               iddy * ValueType(invScaleSqr[1]) +
+               iddz * ValueType(invScaleSqr[2]);
```

Comment:
I'm guessing we want methods like these to always operate at double precision? I was a bit wary of casting/converting values away *from* grid `ValueType`s due to the mindset behind the `zeroVal` idiom. There also seems to be a bit of inconsistency between the different specialised structs. Using the Gradient methods as an example, the base, non specialised version stores the return of `ISGradient` at double precision, explicitly also doing the calculation at double precision. All specialised versions store the result as `ValueType`, but are implicitly doing some computation at double precision due to the return type of `getInvTwiceScale()`. An example change could be:

```
    result(const UniformScaleMap& map, const Accessor& grid, const Coord& ijk)
    {
        using ValueType = typename internal::ReturnValue<Accessor>::ValueType;
        using Vec3Type = typename internal::ReturnValue<Accessor>::Vec3Type;

        Vec3Type iGradient( ISGradient<CD_2NDT>::result(grid, ijk) );
        ValueType inv2dx = ValueType(map.getInvTwiceScale()[0]);
        return  iGradient * inv2dx;
    }
```

to

```
    result(const UniformScaleMap& map, const Accessor& grid, const Coord& ijk)
    {
        using Vec3Type = typename internal::ReturnValue<Accessor>::Vec3Type;

        Vec3d iGradient( ISGradient<CD_2NDT>::result(grid, ijk) );
        double inv2dx = map.getInvTwiceScale()[0];
        return  Vec3Type(iGradient * inv2dx);
    }
```

But this would then set the precedent to change **all** operations to double precision. Methods in interpolation also suffer from this (computation at `ValueType` but with a double precision position coef).

Reply:
I think if we update `Vec3<T>::operator*()` to use `math::promote<>` internally, then the above could reduce to
```
result(const UniformScaleMap& map, const Accessor& grid, const Coord& ijk)
{
    using Vec3Type = typename internal::ReturnValue<Accessor>::Vec3Type;
    const double inv2dx = map.getInvTwiceScale()[0];
    return Vec3Type(ISGradient<CD_2NDT>::result(grid, ijk) * inv2dx);
}
```

Reply:
I think I'd prefer the precision of the calculations to be explicit rather than inferred in the operators i.e. relying on `promote` from the Vec3 op? Also, do we want to add automatic type conversion to math operators? Based on comments above, I was thinking that `Vec3i(1) * 1.0f` should give you the same conversion warning as `1 * 1.0f`

Reply:
Yes, I suppose that's right.  But then I guess for consistency none of the `Vec` operators should do any type conversion, meaning that a few, including `Vec3::operator*=(scalar)`, would need to be changed.